Decoupled the GQL API from the database, so queries are lazily evaluated

Infirmarian opened the pull request:
Modified the GQL schema to represent circular relationships, enabled by the decoupling of the API
closes #25

Infirmarian committed:
Decoupled the GQL API from the database, so queries are lazily evaluated
Modified the GQL schema to represent circular relationships, enabled by the decoupling of the API

Infirmarian committed:
Added back creatorId to art, allowing for better fetching
Rebuilt schemas

Timthetic commented on server/src/graphql/api.ts:
```diff
@@ -72,15 +62,17 @@ export const graphqlRoot: Resolvers<Context> = {
-        })
-      }
-      return result
+      // if (checkSeen) {
+      //   // This statement requires a semicolon
+      //   const seenArt: Set<number> = new Set()
+      //   ;(await User.findOne())?.artSeen.forEach(art => {
+      //     seenArt.add(art.id)
+      //   })
+      //   result.forEach((art: any) => {
+      //     art.seen = seenArt.has(art.id)
+      //   })
+      // }
+      return result as any
```
It occurs to me that we have `seen` as a boolean field on art on the API level. I think the intent was that this was populated using the user context of the request (so we don't get it for free from the DB) so we can color the points on the map. Do we still want to do this, or simply grab the user (and its seen art) on the front end and process it there?

euglo replied:
If it was the way before, I'd only need the single query to fetch nearby artwork and check the `seen` boolean flag for map coloring, but I think with the way it is now I'll have to do queries for both artworkSeen and nearby, then filter the artwork list in the frontend. I think it'll be more efficient the first way, but lmk if you disagree

Timthetic replied:
In the DB, I guess we could add a seenBy field on art, then just use that? I'm kind of surprised that I didn't put the field in previously.

For clarification, the corresponding field already exists on User.

Infirmarian replied:
I got rid of this because the idea of passing a parameter and querying GraphQL that way is definitely an anti-pattern. I guess the best way to deal with this in an efficient manner would be to write our query to get a single piece of artwork just
```sql
SELECT * FROM art WHERE id = '1';
```
Then we can have separate resolver for the `seenByUser` and `likedByUser` fields do something like
```sql
SELECT 1 FROM artSeenJoinTable WHERE userid = x AND artid = y;
```
That could be efficiently rewritten into a dataloader-type request like
```sql
SELECT id, CASE WHEN artid IS NULL FALSE ELSE TRUE
FROM artid
LEFT JOIN artSeenTable t ON t.artid = id 
WHERE id IN (...) AND userid = x;
```

Infirmarian committed:
Added fields for artSeen, artLiked for user
Moved join table ownership for `views` and `likes` to Art
Got a (reasonably) efficient query for `seen` on Art
Cleaned up some code

Infirmarian committed:
Merge branch 'master' into db-decouple
Regenerated query.gen.ts, frontend still compiles ok

euglo commented on server/src/graphql/api.ts:
```diff
@@ -62,16 +68,6 @@ export const graphqlRoot: Resolvers<Context> = {
         })
         .limit(25)
```
I removed this line so that all artwork is returned, since we didn't implement pagination

euglo commented on server/src/graphql/api.ts:
```diff
@@ -24,22 +24,29 @@ interface Context {
-      return response as any
-    },
+    art: async (_, { id }) => Art.findOne({ where: { id: id } }) as any,
+    arts: async () => Art.find({ take: 128 }) as any,
```
Removed take: 128 on my branch

Infirmarian replied:
I'm going to leave this in, since we never query arts on the frontend (as far as I can see), and this way debugging with a large number of arts will be efficient

Infirmarian committed:
Merge branch 'master' into db-decouple
Modified some queries on the frontend to work with the datamodel
